File: y2024_C#/AdventOfCode/Day_05/DayFive.cs
using AdventofCode._General;

namespace AdventOfCode.Day_05;

public class DayFive : Day {
    private readonly List<Rule> _rules;
    private readonly List<Update> _updates;


    public DayFive() : base("input_05.txt") {
        var parser = new Parser(Input);
        _rules = parser.GetRules();
        _updates = parser.GetUpdates();
    }

    public override int PartOne() {
        SumUpCompliantPages();

        return ResultPartOne;
    }

    private void SumUpCompliantPages() {
        var sum = _updates
            .Select(update => _rules
                    .All(rule => rule.Allows(update))
                    ? update.MiddlePage
                    : 0
            ).Sum();

        ResultPartOne = sum;
    }

    public override int PartTwo() {
        SortWrongUpdatesAndSumUpPages();

        return ResultPartTwo;
    }

    private void SortWrongUpdatesAndSumUpPages() {
        var sum = 0;
        foreach (var update in _updates) {
            var gotSorted = SortIfNecessary(update);
            if (gotSorted) {
                sum += update.MiddlePage;
            }
        }

        ResultPartTwo = sum;
    }

    private bool SortIfNecessary(Update update) {
        var concernedRules = update.GetConcernedRules(_rules);
        var compliant = update.CompliantWith(concernedRules);

        if (compliant) {
            return false;
        }

        update.Sort(concernedRules);
        return true;
    }
}

File: y2025_Java/src/day05/DayFive.java
package day05;

import _general.Day;

public class DayFive extends Day {
    private final Inventory inventory;

    public DayFive() {
        this.inventory = new Inventory(input);
    }

    public static void main(String[] args) {
        var day = new DayFive();
        day.printResults();
    }

    @Override
    protected String partOne() {
        var count = inventory.countFreshAvailableIngredients();
        return String.valueOf(count);
    }

    @Override
    protected String partTwo() {
        var sum = inventory.sumUpRangeSizes();
        return String.valueOf(sum);
    }
}
